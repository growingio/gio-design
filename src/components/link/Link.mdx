import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Link from './index';

# Link 链接

<Subtitle>用于表示当前内容可进行页面跳转。</Subtitle>

## 参数说明

<ArgsTable of={Link} />

## 代码演示

### 默认

默认使用 a 标签作为组件。

[Example](/?path=/story/basic-components-link--default)

### 禁用

失效状态，不可点击。

[Example](/?path=/story/basic-components-link--disabled)

### 自定义组件

可以 component props 来自定义组件。

[Example](/?path=/story/basic-components-link--custom-component)
