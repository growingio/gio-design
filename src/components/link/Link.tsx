import React from 'react';
import classnames from 'classnames';
import { LinkProps } from './interface';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

export { LinkProps } from './interface';

const Link: React.FC<LinkProps> = ({
  component = 'a',
  to = '',
  disabled,
  prefix: customPrefixCls,
  className,
  children,
  ...otherProps
}: LinkProps) => {
  const prefixCls = usePrefixCls('link', customPrefixCls);
  const cls = classnames(className, prefixCls, {
    [`${prefixCls}--disabled`]: disabled,
  });

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
    if (disabled) {
      e.preventDefault();
      return;
    }

    if (otherProps.onClick) {
      otherProps.onClick(e);
      return;
    }

    if (to) {
      window.location.href = to;
    }
  };

  const componentProps = {
    className: cls,
    ...otherProps,
    onClick: handleClick,
  };

  if (component === 'a') {
    return (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <a href={to} {...componentProps}>
        {children}
      </a>
    );
  }

  const ComponentProp = component;
  return <ComponentProp {...componentProps}>{children}</ComponentProp>;
};

export default Link;
