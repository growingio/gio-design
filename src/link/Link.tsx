import { LoadingOutlined } from '@gio-design/icons';
import classNames from 'classnames';
import React from 'react';
import usePrefixCls from '../utils/hooks/use-prefix-cls';
import { LinkProps } from './interface';

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const { className, children, prefix, loading = false, disabled = false, ...restProps } = props;

  const prefixCls = usePrefixCls('link');
  const classes = classNames([prefixCls, className], {
    [`${prefixCls}_disabled`]: disabled,
    [`${prefixCls}_loading`]: loading,
  });

  const prefixIcon = loading ? (
    <span className={`${prefixCls}-prefix-icon`}>
      <LoadingOutlined rotating />
    </span>
  ) : (
    prefix && <span className={`${prefixCls}-prefix-icon`}>{prefix}</span>
  );

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <a className={classes} ref={ref} aria-disabled={disabled || loading} {...restProps}>
      {prefixIcon}
      {children}
    </a>
  );
});

Link.displayName = 'Link';

Link.defaultProps = {
  loading: false,
  disabled: false,
};

export default Link;
