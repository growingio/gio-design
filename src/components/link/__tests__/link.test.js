import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Link from '..';

describe('Test Link', () => {
  const saveLocation = window.location;
  afterAll(() => {
    delete global.window.location;
    global.window.location = saveLocation;
  });

  it('should mount and unmount Link with no error.', () => {
    const wrapper = mount(<Link to="https://www.growingio.com">GrowingIO</Link>);
    expect(() => {
      wrapper.setProps({});
      wrapper.unmount();
    }).not.toThrow();
  });

  it('should render correctly', () => {
    const domTree = renderer
      .create(
        <Link className="gio-customized-link" to="https://www.growingio.com" disabled>
          GrowingIO
        </Link>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  it('should render with custom component `div`', () => {
    const wrapper = shallow(
      <Link to="https://www.growingio.com" component="div">
        Custom Component
      </Link>
    );
    expect(wrapper.name()).toEqual('div');
  });

  it('should response click event', () => {
    const onClick = jest.fn();

    const wrapper = mount(
      <Link to="https://www.growingio.com" component="div" onClick={onClick}>
        Custom Component
      </Link>
    );

    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  it('It should show or hide', () => {
    const wrapper = mount(
      <Link component="div" to="https://www.growingio.com" disabled>
        GrowingIO
      </Link>
    );
    wrapper.simulate('click');
    expect(wrapper.exists('.gio-link--disabled')).toBe(true);
  });
  it('should accept undefined link', () => {
    const onClick = jest.fn();

    const wrapper = mount(
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <Link component="div" to={undefined} onClick={onClick}>
        GrowingIO
      </Link>
    );
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  it('should direct to right location', () => {
    delete global.window.location;
    global.window.location = { href: '' };

    const wrapper = mount(
      <Link component="div" to="https://www.growingio.com">
        GrowingIO
      </Link>
    );
    wrapper.simulate('click');
    expect(window.location.href).toEqual('https://www.growingio.com');
  });
  it('not pointing in the right place', () => {
    delete global.window.location;
    global.window.location = { href: '' };

    const wrapper = mount(
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      <Link component="div" to={undefined}>
        GrowingIO
      </Link>
    );
    wrapper.simulate('click');
    expect(window.location.href).not.toEqual('https://www.growingio.com');
  });
});
