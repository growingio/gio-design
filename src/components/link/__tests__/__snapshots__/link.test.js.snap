// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Link should render correctly 1`] = `
<a
  className="gio-customized-link gio-link gio-link--disabled"
  href="https://www.growingio.com"
  onClick={[Function]}
>
  GrowingIO
</a>
`;
